[Search git status] preview changed files

NextAlone opened the pull request:
Add a preview for the changed file

PatrickF1 commented:
Don't let the fact the other two PRs didn't get merged discourage you. It was more because the authors lost interest than this feature being impossible to work on (I hope).
If you can make the preview show both the staged and working directory diffs against HEAD, then I am happy to merge.
Two potential solutions:
- run git diff and git diff —index separately and append them to a single file, then cat it
https://stackoverflow.com/questions/67150991/show-both-worktree-and-index-changes-for-a-single-path-in-git
- use git status vv

NextAlone commented:
> Don't let the fact the other two PRs didn't get merged discourage you. It was more because the authors lost interest than this feature being impossible to work on (I hope).
> If you can make the preview show both the staged and working directory diffs against HEAD, then I am happy to merge.
> Two potential solutions:
> - run git diff and git diff —index separately and append them to a single file, then cat it
> https://stackoverflow.com/questions/67150991/show-both-worktree-and-index-changes-for-a-single-path-in-git
> - use git status vv

Ok, i will check these two.

NextAlone commented:
> Don't let the fact the other two PRs didn't get merged discourage you. It was more because the authors lost interest than this feature being impossible to work on (I hope). If you can make the preview show both the staged and working directory diffs against HEAD, then I am happy to merge. Two potential solutions:
> 
> * run git diff and git diff —index separately and append them to a single file, then cat it
>   https://stackoverflow.com/questions/67150991/show-both-worktree-and-index-changes-for-a-single-path-in-git
> * use git status vv

Are you talking about being displayed together or separately?

NextAlone commented:
How about this?

```fish
if string match -r \'\?\?\' {1} > /dev/null
    _fzf_preview_file {2..}
else if string match -r \'A\' {1} > /dev/null
    git diff --color=always -- /dev/null {2..}
else
    git diff --color=always HEAD {2..}
end'
```

PatrickF1 commented:
> Are you talking about being displayed together or separately?

If there are both staged and working tree changes, I want the preview to show both.

Yeah the second one is along the lines of what I'm thinking. I guess you don't need the intermediate file to store them huh?

NextAlone commented:
```
 X          Y     Meaning
 -------------------------------------------------
          [AMD]   not updated
 M        [ MTD]  updated in index
 T        [ MTD]  type changed in index
 A        [ MTD]  added to index
 D                deleted from index
 R        [ MTD]  renamed in index
 C        [ MTD]  copied in index
 [MTARC]          index and work tree matches
 [ MTARC]    M    work tree changed since index
 [ MTARC]    T    type changed in work tree since index
 [ MTARC]    D    deleted in work tree
             R    renamed in work tree
             C    copied in work tree
 -------------------------------------------------
 D           D    unmerged, both deleted
 A           U    unmerged, added by us
 U           D    unmerged, deleted by them
 U           A    unmerged, added by them
 D           U    unmerged, deleted by us
 A           A    unmerged, both added
 U           U    unmerged, both modified
 -------------------------------------------------
 ?           ?    untracked
 !           !    ignored
 -------------------------------------------------
```
Not works with R(renamed), C(copied), T(type changed), and unmerged status. i will find out the way.

NextAlone commented:
> > > Are you talking about being displayed together or separately?
> > 
> > 
> > If there are both staged and working tree changes, I want the preview to show both.
> > Yeah the second one is along the lines of what I'm thinking. I guess you don't need the intermediate file to store them huh?
> 
> Yes, no intermediate file, and add check with status (just simply string match, don't know other convenient way)?
> 
> ```fish
> if string match -r '\?\?' {1} > /dev/null
>     _fzf_preview_file {2..}
> else if string match -r 'A' {1} > /dev/null
>     git diff --color=always -- /dev/null {2..}
> else 
>     if string match -r '\S\s\S' > /dev/null
>         echo -e (set_color red)Unstaged
>         echo
>         git diff --color=always HEAD -- {2..}
>         echo
>     end
>     if string match -r '\s\s|\S\S\s' > /dev/null
>         echo (set_color green)Staged
>         echo
>         git diff --color=always --cached HEAD -- {2..}
>     end
> end
> ```
> 
> ![image](https://user-images.githubusercontent.com/12210746/167903461-e613b0d4-3c61-460c-9d9d-9f6ae816c340.png)

maybe this is enough to preview

PatrickF1 commented on functions/_fzf_search_git_status.fish:
```diff
@@ -8,6 +8,7 @@ function _fzf_search_git_status --description "Search the output of git status.
             _fzf_wrapper --ansi \
                 --multi \
                 --query=(commandline --current-token) \
+                --preview='set -l path (string split " " {})[-1]; if not string match -r "\?\?.*" {}; if string match -r "^[NMAD]\s\s\S*" {}; git diff --color=always --cached -- $path; else if string match -r "^\s[NMAD]\s\S*" {}; git diff --color=always -- $path; else if string match -r "^[NMAD][NMAD]\s\S*" {}; git diff --color=always -- $path; git diff --color=always --cached -- $path; end; end' ...
```
Can you make this a helper function so it's easier to review and maintain? see the preview function for other features for examples

PatrickF1 commented on functions/_fzf_preview_status.fish:
```diff
@@ -0,0 +1,24 @@
+function _fzf_preview_status
```
can you renamed to changed file?

NextAlone replied:
you mean `_fzf_preview_changed_file` ?

PatrickF1 replied:
yeah, sorry I wasn't clear

PatrickF1 commented on functions/_fzf_preview_status.fish:
```diff
@@ -0,0 +1,24 @@
+function _fzf_preview_status
+    set -l path (string split ' ' $argv)[-1]
+    if string match -r '^\?\?' $argv >/dev/null
```
having a bit of trouble following the logic of each if/else if/else. could you outline the "algorithm" at a high level and explain why you think it's comprehensive?

NextAlone replied:
1. use `_fzf_preview_file` to preview untracked files or dirs
2. diff with /dev/null and new files
3. then other files are indexed, then seperate unstaged and staged files as you mentioned.

I think that this basically covers common situations

PatrickF1 reviewed (commented):
thanks for the great work so far!

PatrickF1 commented on functions/_fzf_preview_changed_file.fish:
```diff
@@ -0,0 +1,24 @@
+    if string match -r '^\?\?' $argv >/dev/null
+        echo Untracked
+        echo
+        _fzf_preview_file $path
```
niceeeee, so clever to go call _fzf_preview_file
can you go to the top of _fzf_preview_file and remove the comment?

NextAlone replied:
is this? 
`# helper function for _fzf_search_directory`

PatrickF1 replied:
yup, actually can you update it to say `helper function for _fzf_search_directory and _fzf_preview_changed_file`

NextAlone replied:
> yup, actually can you update it to say `helper function for _fzf_search_directory and _fzf_preview_changed_file`

emm, you mean `_fzf_search_directory` and `_fzf_search_git_status` right?

PatrickF1 commented on functions/_fzf_preview_changed_file.fish:
```diff
@@ -0,0 +1,24 @@
+function _fzf_preview_changed_file
+    set -l path (string split ' ' $argv)[-1]
+    if string match -r '^\?\?' $argv >/dev/null
```
you can do --quiet to avoid >/dev/null

PatrickF1 commented on functions/_fzf_preview_changed_file.fish:
```diff
@@ -0,0 +1,24 @@
+        echo Untracked
+        echo
+        _fzf_preview_file $path
+    else if string match -r '^A\s\s\S' $argv >/dev/null
```
For the regex, I get why the ^A, but how come you also have to do \s\s\S?

NextAlone replied:
to match `[AMDCRT][MCDRT] file`

I haven't tried to simplify it since it works.

PatrickF1 commented on functions/_fzf_preview_changed_file.fish:
```diff
@@ -0,0 +1,24 @@
+            git diff --color=always -- $path
+            echo
+        end
+        if string match -r '\S\s\s\S|\S\S\s\S' $argv >/dev/null
```
sorry I know you tried to explain what you were doing but I'm still confused about this regex and the one above

NextAlone replied:
Maybe `^\S` also works.

NextAlone commented:
No need to split `^A`, it works with staged file

NextAlone commented:
> Thanks for the updates. This week is really busy so it'll be a while before I can come back to review but just letting you know I won't forget about this

It doesn't matter, I will always be here.

NextAlone commented:
> Thanks for the updates. This week is really busy so it'll be a while before I can come back to review but just letting you know I won't forget about this

It doesn't matter, I will always be here.

PatrickF1 commented:
Don't worry I haven't forgotten about this, just thinking through what is the best way to make this feature simple and easy to use!

PatrickF1 commented on functions/_fzf_preview_changed_file.fish:
```diff
@@ -0,0 +1,20 @@
+function _fzf_preview_changed_file
+    set -l path (string split ' ' $argv)[-1]
+    if string match -r '^\?\?' $argv --quiet
```
Hi, I'm back! I've found a great way to extract the symbols without using regex--you can use `git status --porcelain`. Can you update the code to use that instead? Afterwards, we'll be very close to merging

NextAlone replied:
I didn't quite understand it and didn't find its use effect

![image](https://user-images.githubusercontent.com/12210746/172096898-56828210-8c71-4c58-86d4-23316742486a.png)

PatrickF1 replied:
Try using version 1 porcelain and qualifying the file path. The status it displays is much easier to parse because it's at the very beginning of the line. Then you can use substring instead of playing with unreadable regexes.

NextAlone replied:
> 

but it still needs to regex with prefix to detect file status

PatrickF1 replied:
What I'd strongly prefer to see is something like this pattern
```fish
set file_status (command to grab the two letters of the file status)
if test file_status[0] = "A"

```
I think this'll be much more readable. Right now I still don't exactly know what your regexes are matching.

NextAlone replied:
> What I'd strongly prefer to see is something like this pattern
> 
> ```fish
> set file_status (command to grab the two letters of the file status)
> if test file_status[0] = "A"
> ```
> 
> I think this'll be much more readable. Right now I still don't exactly know what your regexes are matching.

`--porcelain` no diff to `-s`
![image](https://user-images.githubusercontent.com/12210746/173193794-f14c3e09-e22b-4dfb-9831-5802ed9a1c5e.png)

NextAlone replied:
> What I'd strongly prefer to see is something like this pattern
> 
> ```fish
> set file_status (command to grab the two letters of the file status)
> if test file_status[0] = "A"
> ```
> 
> I think this'll be much more readable. Right now I still don't exactly know what your regexes are matching.

`git status` shows with two character, when first not null, it means staged changes, while second, the unstaged.

So, `\S\s\S` means unstaged and `^\S` means staged

PatrickF1 replied:
Got it, I get it now, thanks! I am fine with the logic as it is then.

NextAlone commented:
And if there any other changes request？

PatrickF1 commented:
Nope, I'll take it from here and merge soon :)

PatrickF1 commented:
Thanks for working on this and especially our patience @NextAlone! Really appreciate it. You are the third person to try to PR this change in and it seems third time's the charm. Sorry I've been so busy.

NextAlone commented:
> Thanks for working on this and especially our patience @NextAlone! Really appreciate it. You are the third person to try to PR this change in and it seems third time's the charm. Sorry I've been so busy.

Also thanks for your work, this is such a wonderful plugin.

NextAlone commented:
> Thanks for working on this and especially our patience @NextAlone! Really appreciate it. You are the third person to try to PR this change in and it seems third time's the charm. Sorry I've been so busy.

Also thanks for your work, this is such a wonderful plugin.

davidnanry commented:
Thank you both! This is an excellent addition to an excellent plugin

NextAlone commented:
@PatrickF1 And it is time to update `Modified paths` section in Readme.

PatrickF1 commented:
I did update it. I'll add a gif later. Or did you mean something else?
